virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+
+1. extra-terrestrial radiation (affected by the earth's orbit, date, and location),
+2. terrestrial radiation (affected by atmospheric composition and clouds),
+3. topography (elevation, slope and aspect),
+4. surface albedo (vegetation type and fraction of vegetation/bare soil), and
+5. emitted longwave radiation.
+
+The preprocessing module takes extra-terrestrial radiation as an input and adjusts for
+the effects of topography (slope and aspect). Here, the effects of atmospheric
+filtering (elevation-dependent) and cloud cover are added to calculate photosynthetic
+photon flux density (PPFD) at the top of the canopy which is a crucial input to the
+plant module. The implementation is based on :cite:t:`Davis2017`.
+
+Cloud cover and surface albedo also determine how much of the shortwave radiation that
+reaches the top of the canopy is reflected and how much remains to be absorbed via
+photosynthesis and re-emitted as longwave radiation by vegetation and forest floor.
+At this stage, scattering and re-absorption of longwave radiation are not considered.
+
+"""
+# the following structural components are not implemented yet
+# TODO include time dimension
+# TODO logging, raise errors
+
+
+from dataclasses import dataclass, field
+from typing import Any, Dict
+
+import numpy as np
+from numpy.typing import NDArray
+
+from virtual_rainforest.core.logger import LOGGER
+from virtual_rainforest.core.model import InitialisationError
+
+# from some external source, could be core.constants or abiotic.config
+CLOUDY_TRANSMISSIVITY = 0.25
+"""Cloudy transmittivity :cite:p:`Linacre1968`"""
+TRANSMISSIVITY_COEFFICIENT = 0.50
+"""Angular coefficient of transmittivity :cite:p:`Linacre1968`"""
+FLUX_TO_ENERGY = 2.04
+"""From flux to energy conversion, umol J-1 :cite:p:`Meek1984`"""
+STEFAN_BOLTZMANN_CONSTANT = 5.67e-8
+"""Stefan-Boltzmann constant W m-2 K-4"""
+SOIL_EMISSIVITY = 0.95
+"""Soil emissivity, default for tropical rainforest"""
+CANOPY_EMISSIVITY = 0.95
+"""Canopy emissivity, default for tropical rainforest"""
+BEER_REGRESSION = 2.67e-5
+"""Parameter in equation for atmospheric transmissivity based on regression of Beer's
+radiation extinction function :cite:p:`Allen1996`"""
+CELSIUS_TO_KELVIN = 273.15
+"""Factor to convert temperature in Celsius to absolute temperature in Kelvin"""
+
+# dummy data object, 2 grid cells, 3 canopy layers
+data: Dict[str, Any] = {
+    "elevation": np.array([100, 100], dtype=np.float32),
+    "sunshine_fraction": np.array([0.5, 1.0], dtype=np.float32),
+    "shortwave_in": np.array([100, 100], dtype=np.float32),
+    "canopy_temperature": np.array(
+        [
+            [25, 22, 20],
+            [25, 22, 20],
+        ],
+        dtype=np.float32,
+    ),
+    "surface_temperature": np.array([20, 20], dtype=np.float32),
+    "canopy_absorption": np.array(
+        [
+            [300, 200, 100],
+            [300, 200, 100],
+        ],
+        dtype=np.float32,
+    ),
+}
```

Comment:
Now that the `data` branch is in, we can start using the actual data system 😄 The code below creates an actual `Data` object with the same contents:

```python
from virtual_rainforest.core.data import Data
from virtual_rainforest.core.grid import Grid
from xarray import DataArray


grid = Grid(cell_nx=2, cell_ny=1)
data = Data(grid)

data['elevation'] = DataArray([100, 100], dims=['cell_id'])
data['sunshine_fraction'] = DataArray([0.5, 1.0], dims=['cell_id'])
data['canopy_temperature'] = DataArray([
            [25, 22, 20],
            [25, 22, 20],
        ],
        dims = ['cell_id', 'canopy_layer']
    )
data['surface_temperature'] = DataArray([20, 20], dims=['cell_id'])
data['canopy_absorption'] = DataArray([
            [300, 200, 100],
            [300, 200, 100],
        ],
        dims = ['cell_id', 'canopy_layer']
    )

```